standard/variables.md

```diff
@@ -901,3 +901,181 @@ variable_reference
+
+- The *caller-scope*.
+- The *ref-safe-scope* of all `ref` and `out` argument expressions (excluding the receiver).
+- For each `in` parameter of the method, if there is a corresponding expression that is an variable, its *ref-safe-scope*, otherwise the nearest enclosing scope
+- The *safe-scope* of all argument expressions (including the receiver)
+
+> *Example*: the last bullet is necessary to handle code such as
+>
+> ```csharp
+> var sp = new Span(...)
+> return ref sp[0];
+> ```
+>
+> or
+>
+> ```csharp
+> return ref M(sp, 0);
+> ```
+>
+> *end example*
+
+For a value resulting from a method invocation `e1.M(e2, ...)`, its *safe-scope* from the narrowest of the following scopes:
+
+- The *caller-scope*
+- The *safe-scope* of all argument expressions (including the receiver)
+
+A property invocation (either `get` or `set`) is treated as a method invocation of the underlying method by the above rules.
+
+### §ref-span-safety-a-value Values
+
+A value's *ref-safe-scope* is the nearest enclosing scope.
+
+> *Note:* This occurs in an invocation such as `M(ref d.Length)` where `d` is of type `dynamic`. It is also consistent with arguments corresponding to `in` parameters.
+
+### §ref-span-safety-stackalloc stackalloc
+
+The *safe-scope* for a value `c` resulting from a `stackalloc` expression is the current method.
+
+#### §ref-span-safety-constructor-invocations Constructor invocations
+
+A `new` expression that invokes a constructor obeys the same rules as a method invocation (§ref-span-safety-method-invocation) that is considered to return the type being constructed.
+
+In addition, for a value `c` that is the result of a `new` expression, the *safe-scope* is no wider than the smallest of the *safe-to-escape-scopes* of all arguments and operands of the object initializer expressions, recursively, if any initializer is present.
+
+### §ref-span-safety-default-expressions Default expressions
+
+The *safe-scope* of a `default` expression is the *caller-scope*.
+
+### §ref-span-safety-limitations Limitations on reference variables
+
+- Neither a `ref` parameter, nor a `ref` local, nor a parameter or local of a `ref struct` type can be lifted into a lambda expression or local function.
+- Neither a `ref` parameter nor a parameter of a `ref struct` type may be an argument for an iterator method or an `async` method.
+- Neither a `ref` local, nor a local of a `ref struct` type may be in scope at the point of a `yield return` statement or an `await` expression.
+- A `ref struct` type may not be used as a type argument, or as an element type in a tuple type.
+- A `ref struct` type may not be the declared type of a field, except that it may be the declared type of an instance field of another `ref struct`.
+- A `ref struct` type may not be the element type of an array.
+- A value of a `ref struct` type may not be boxed:
```

Comment:
We probably need to modify 10.2.9 for this. (It may be in your follow-up list already.) Various other restrictions below probably need propagating as well.

Reply:
I'd like to defer these references to these PRs:

- #44 out variables (which may just be closed now)
- #213 ref locals and returns
- #219 `in` parameters
- #601 `ref struct` types

I think the text should be in the section that defines each of those variable types. These rules may be removed at that point.

Reply:
Action from 2023-04-19: agree to defer.